File: character_logic/ashend.py
"""Ashend battle and AI hooks."""

from __future__ import annotations

from typing import Any


CHARACTER_ID = "ashend"
HIDDEN_COUNTERS = {"재로부터의 엄습", "재가 되어 회피"}


def extra_state_parts(battle: Any, fighter: Any) -> list[str]:
    parts: list[str] = []
    if fighter.counters.get("재로부터의 엄습", 0) > 0:
        parts.append("다음 공격 피해 x1.5")
    if fighter.counters.get("재가 되어 회피", 0) > 0:
        parts.append(f"공격 회피 50% · {fighter.counters['재가 되어 회피']}턴")
    return parts


def reset_turn_flags(battle: Any, fighter: Any) -> None:
    if fighter.counters.get("재가 되어 회피", 0) > 0:
        fighter.evasion_chance += 50


def modify_accuracy_status(battle: Any, choice: Any, target: Any, accuracy: float) -> float:
    if "회진" in choice.actor.statuses and choice.action.is_attack:
        accuracy *= 0.8
    return accuracy


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    return None


def modify_attack_power(battle: Any, choice: Any, power: int) -> int:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 2):
        if battle.record.attack_damage_taken.get(actor.side, 0) <= 0 and "회진" in target.statuses:
            turns = target.statuses["회진"].remaining
            choice.power = (choice.power or 0) + turns * 8
            power = int(choice.power)
    return power


def estimated_power(battle: Any, actor: Any, target: Any, action: Any, power: int) -> int:
    if action.is_skill(CHARACTER_ID, 2):
        if battle.record.attack_damage_taken.get(actor.side, 0) <= 0 and "회진" in target.statuses:
            power += target.statuses["회진"].remaining * 8
    return power


def attack_damage_multipliers(battle: Any, choice: Any) -> list[float]:
    if choice.actor.counters.get("재로부터의 엄습", 0) > 0:
        return [1.5]
    return []


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0):
        turns = battle.rng.randint(2, 5)
        print(f"회진 지속시간 {turns}턴으로 결정되었다.")
        battle.add_status(target, "회진", turns, 1, actor.name)
    elif action.is_skill(CHARACTER_ID, 2):
        battle.add_status(target, "회진", 2, 1, actor.name)


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if action.is_skill(CHARACTER_ID, 1):
        actor.guaranteed_evasion = True
        print("이번 턴 이후 상대 공격에 대한 회피 판정이 반드시 성공한다.")
        return True
    if action.is_skill(CHARACTER_ID, 3):
        actor.evasion_chance += 50
        actor.counters["재가 되어 회피"] = 4
        print("4턴 동안 상대의 공격을 50% 확률로 회피한다.")
        return True
    return False


def on_turn_end(battle: Any, fighter: Any) -> None:
    opponent = battle.opponent(fighter)
    if battle.kind_is_attack(battle.record.selected_kind.get(opponent.side)) and battle.record.attack_damage_taken.get(fighter.side, 0) <= 0:
        fighter.counters["재로부터의 엄습"] = 2
        print(f"{fighter.name}은 다음 턴 공격 피해가 1.5배가 된다.")
    if battle.record.selected_key.get(fighter.side) == battle.skill_key(CHARACTER_ID, 1):
        if battle.kind_is_attack(battle.record.selected_kind.get(opponent.side)) and battle.record.attack_damage_taken.get(fighter.side, 0) <= 0:
            battle.add_stat_effect(fighter, "atk", 1.4, 4, "회색의 안개 속으로")


def decrement_counters(fighter: Any) -> None:
    if fighter.counters.get("재로부터의 엄습", 0) > 0:
        fighter.counters["재로부터의 엄습"] -= 1
    if fighter.counters.get("재가 되어 회피", 0) > 0:
        fighter.counters["재가 되어 회피"] -= 1


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    dust = target.statuses.get("회진")
    dust_remaining = dust.remaining if dust else 0
    incoming = battle.estimate_best_incoming_damage(target, actor)
    counts = battle.recent_kind_counts(target)
    attack_read = counts["attack"]
    ambush_ready = actor.counters.get("재로부터의 엄습", 0) > 0

    if ambush_ready and action.is_attack:
        value += expected_damage * 0.7 + 260
    if action.is_skill(CHARACTER_ID, 0):
        if dust is None:
            value += 540 * hit_rate
        elif dust_remaining <= 1:
            value += 220
        else:
            value -= 120
    if action.is_skill(CHARACTER_ID, 1):
        value += incoming * 2.1 + attack_read * 420
        if incoming > 0:
            value += 760
        if ambush_ready:
            value -= 220
    if action.is_skill(CHARACTER_ID, 2):
        if dust is not None:
            value += dust_remaining * 340 + expected_damage * 1.1
            if expected_damage >= target.hp:
                value += 2600
        else:
            value -= 260
        if incoming > actor.hp * 0.25 and attack_read > 0:
            value -= 250
    if action.is_skill(CHARACTER_ID, 3):
        value += incoming * 1.35 + attack_read * 260
        if actor.counters.get("재가 되어 회피", 0) <= 0:
            value += 620
        if actor.hp <= incoming:
            value += 480
    return value

File: web/battle-engine/character-logic/ashend.js
"use strict";

const CHARACTER_ID = "ashend";

function log(battle, message) {
  battle.logs.push(message);
}

function recentKindCounts(battle, fighter) {
  return typeof battle.recentKindCounts === "function"
    ? battle.recentKindCounts(fighter)
    : { attack: 0, defense: 0, meditation: 0 };
}

function modifyAccuracyStatus(_battle, choice, _target, accuracy) {
  if (choice.actor.statuses["회진"] && choice.action.isAttack) return accuracy * 0.8;
  return accuracy;
}

module.exports = {
  extraStateParts(_battle, fighter) {
    const parts = [];
    if (Number(fighter.counters["재로부터의 엄습"] || 0) > 0) parts.push("다음 공격 피해 x1.5");
    if (Number(fighter.counters["재가 되어 회피"] || 0) > 0) {
      parts.push(`공격 회피 50% · ${fighter.counters["재가 되어 회피"]}턴`);
    }
    return parts;
  },

  resetTurnFlags(_battle, fighter) {
    if (Number(fighter.counters["재가 되어 회피"] || 0) > 0) fighter.evasionChance += 50;
  },

  modifyAccuracyStatus,

  applyConditionEffects() {
    return true;
  },

  modifyAttackPower(battle, choice, power) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    if (
      choice.action.isSkill(CHARACTER_ID, 2)
      && Number(battle.record.attackDamageTaken[actor.side] || 0) <= 0
      && target.statuses["회진"]
    ) {
      choice.power = Number(choice.power || 0) + Number(target.statuses["회진"].remaining || 0) * 8;
      return Math.trunc(choice.power);
    }
    return power;
  },

  estimatedPower(battle, actor, target, action, power) {
    if (
      action.isSkill(CHARACTER_ID, 2)
      && Number(battle.record.attackDamageTaken[actor.side] || 0) <= 0
      && target.statuses["회진"]
    ) {
      return power + Number(target.statuses["회진"].remaining || 0) * 8;
    }
    return power;
  },

  attackDamageMultipliers(_battle, choice) {
    return Number(choice.actor.counters["재로부터의 엄습"] || 0) > 0 ? [1.5] : [];
  },

  onHitAfterDefenseAsActor(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    if (choice.action.isSkill(CHARACTER_ID, 0)) {
      const turns = 2 + battle.rng.range(4);
      log(battle, `회진 지속시간 ${turns}턴으로 결정되었다.`);
      battle.addStatus(target, "회진", turns, 1, actor.name);
    } else if (choice.action.isSkill(CHARACTER_ID, 2)) {
      battle.addStatus(target, "회진", 2, 1, actor.name);
    }
  },

  applyNonAttackEffects(battle, choice) {
    const actor = choice.actor;
    if (choice.action.isSkill(CHARACTER_ID, 1)) {
      actor.guaranteedEvasion = true;
      log(battle, "이번 턴 이후 상대 공격에 대한 회피 판정이 반드시 성공한다.");
      return true;
    }
    if (choice.action.isSkill(CHARACTER_ID, 3)) {
      actor.evasionChance += 50;
      actor.counters["재가 되어 회피"] = 4;
      log(battle, "4턴 동안 상대의 공격을 50% 확률로 회피한다.");
      return true;
    }
    return false;
  },

  onTurnEnd(battle, fighter) {
    const opponent = battle.opponent(fighter);
    const opponentAttacked = battle.kindIsAttack(battle.record.selectedKind[opponent.side]);
    const avoidedDamage = Number(battle.record.attackDamageTaken[fighter.side] || 0) <= 0;
    if (opponentAttacked && avoidedDamage) {
      fighter.counters["재로부터의 엄습"] = 2;
      log(battle, `${fighter.name}은 다음 턴 공격 피해가 1.5배가 된다.`);
    }
    if (battle.record.selectedKey[fighter.side] === `${CHARACTER_ID}:1` && opponentAttacked && avoidedDamage) {
      battle.addStatEffect(fighter, "atk", 1.4, 4, "회색의 안개 속으로");
    }
  },

  decrementCounters(fighter) {
    if (Number(fighter.counters["재로부터의 엄습"] || 0) > 0) fighter.counters["재로부터의 엄습"] -= 1;
    if (Number(fighter.counters["재가 되어 회피"] || 0) > 0) fighter.counters["재가 되어 회피"] -= 1;
  },

  aiScore(battle, actor, target, action, expectedDamage, hitRate) {
    let value = 0;
    const dust = target.statuses["회진"];
    const dustRemaining = dust ? Number(dust.remaining || 0) : 0;
    const incoming = battle.estimateBestIncomingDamage(target, actor);
    const counts = recentKindCounts(battle, target);
    const attackRead = Number(counts.attack || 0);
    const ambushReady = Number(actor.counters["재로부터의 엄습"] || 0) > 0;
    if (ambushReady && action.isAttack) value += expectedDamage * 0.7 + 260;
    if (action.isSkill(CHARACTER_ID, 0)) {
      value += dust ? (dustRemaining <= 1 ? 220 : -120) : 540 * hitRate;
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      value += incoming * 2.1 + attackRead * 420;
      if (incoming > 0) value += 760;
      if (attackRead >= 2 && Number(actor.counters["재가 되어 회피"] || 0) <= 0) value += 620;
      if (ambushReady) value -= 220;
    } else if (action.isSkill(CHARACTER_ID, 2)) {
      if (dust) {
        value += dustRemaining * 340 + expectedDamage * 1.1;
        if (ambushReady) value += 900 + expectedDamage * 0.8;
        if (expectedDamage >= target.hp) value += 2600;
      } else value -= 260;
      if (incoming > actor.hp * 0.25 && attackRead > 0) value -= 250;
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      value += incoming * 1.35 + attackRead * 260;
      if (Number(actor.counters["재가 되어 회피"] || 0) <= 0) value += 620;
      if (actor.hp <= incoming) value += 480;
    }
    return value;
  },
};

module.exports.borrowedEffects = {
  extraStateParts: module.exports.extraStateParts,
  resetTurnFlags: module.exports.resetTurnFlags,
  attackDamageMultipliers: module.exports.attackDamageMultipliers,
  estimatedDamageMultipliers(_battle, actor) {
    return Number(actor.counters["재로부터의 엄습"] || 0) > 0 ? [1.5] : [];
  },
  decrementCounters: module.exports.decrementCounters,
};
